### PR TITLE
Global index cuda

### LINKED ISSUE #560: Support Global Iteration Spaces in the CUDA code-generation
Currently we only support global iteration spaces in the c++-naive back end. This should be extendable to the cuda code generation. 
Pointers to where to change things / where to draw inspiration:
https://github.com/MeteoSwiss-APN/dawn/blob/87969caf9d1c57050cbe3e8aa22788c0ae475f97/dawn/src/dawn/CodeGen/Cuda/MSCodeGen.cpp#L555-L559

https://github.com/MeteoSwiss-APN/dawn/blob/87969caf9d1c57050cbe3e8aa22788c0ae475f97/dawn/src/dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.cpp#L473-L492

An example code for this can be found in 
https://github.com/MeteoSwiss-APN/dawn/blob/master/dawn/examples/python/global_index_stencil.py

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Technical Description

This PR implements support for global iteration spaces in the Cuda backend in addition to the existing support in the naive C++ backend.

### Resolves / Enhances

Resolves #560 

### Testing

The unit test TestCodeGenCuda.GlobalIndexStencil verifies that the Cuda backend generates correct code for a global index example.

### CODE REVIEW

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

File: dawn/test/unit-test/dawn/CodeGen/Cuda/TestCodeGenCuda.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  for(auto const& macroDefine : tu->getPPDefines())
+    ss << macroDefine << "\n";
+
+  ss << tu->getGlobals();
+  for(auto const& s : tu->getStencils())
+    ss << s.second;
+  os << ss.str();
+}
+
+std::string read(const std::string& file) {
```
Comment (twicki) [inline dawn/test/unit-test/dawn/CodeGen/Cuda/TestCodeGenCuda.cpp]: should we export this to `dawn/dawn/src/unittest/` or `dawn/dawn/src/support/` as support functions that are used in unittesting?

File: dawn/src/dawn/CodeGen/Cuda/MSCodeGen.cpp [hunk 2]
```diff
@@ -763,6 +773,21 @@ void MSCodeGen::generateCudaKernelCode() {
     }
   }
 
+  if(iterationSpaceSet_) {
+    for(const auto& stencil : stencilInstantiation_->getStencils()) {
+      for(auto& stage : iterateIIROver<iir::Stage>(*stencil)) {
+        std::string prefix = "int* const stage" + std::to_string(stage->getStageID()) + "Global";
```
Comment (twicki) [inline dawn/src/dawn/CodeGen/Cuda/MSCodeGen.cpp]: can we keep the convention to have west const at least in the generated code?

File: dawn/src/dawn/CodeGen/Cuda/CudaCodeGen.cpp [hunk 3]
```diff
@@ -192,29 +207,34 @@ void CudaCodeGen::generateStencilClasses(
...
 void CudaCodeGen::generateStencilClassMembers(
     Structure& stencilClass, const iir::Stencil& stencil, const sir::GlobalVariableMap& globalsMap,
     IndexRange<const std::map<int, iir::Stencil::FieldInfo>>& nonTempFields,
     IndexRange<const std::map<int, iir::Stencil::FieldInfo>>& tempFields,
-    std::shared_ptr<StencilProperties> stencilProperties) const {
+    std::shared_ptr<StencilProperties> stencilProperties, bool iterationSpaceSet) const {
 
   stencilClass.addComment("Members");
+  if(iterationSpaceSet) {
+    generateGlobalIndices(stencil, stencilClass);
```
Comment (twicki) [inline dawn/src/dawn/CodeGen/Cuda/CudaCodeGen.cpp]: this generates a host function, seems unnecessary, right?

File: dawn/src/dawn/CodeGen/CodeGen.cpp [hunk 4]
```diff
@@ -366,5 +365,59 @@ std::string CodeGen::generateFileName(const stencilInstantiationContext& context
...
+    if(stage->getIterationSpace()[0].has_value()) {
+      stencilClass.addMember("std::array<int, 2>",
+                             "stage" + std::to_string(stage->getStageID()) + "GlobalIIndices");
+    }
+    if(stage->getIterationSpace()[1].has_value()) {
+      stencilClass.addMember("std::array<int, 2>",
+                             "stage" + std::to_string(stage->getStageID()) + "GlobalJIndices");
+    }
+  }
+  // if(iterationSpaceSet) {
```
Comment (twicki) [inline dawn/src/dawn/CodeGen/CodeGen.cpp]: remove comments

File: scripts/ignore_list.sh [hunk 5]
```diff
@@ -3,6 +3,8 @@ ignore_regex_list=(
     "^(\./)?dawn/test/utils/googletest/"
     "^(\./)?dawn/src/dawn/Support/External/"
     "^(\./)?dawn/examples/python/data/"
+    "^(\./)?dawn/test/unit-test/dawn/CodeGen/Naive/reference"
```
Comment (twicki) [inline scripts/ignore_list.sh]: this has been replaced in the mean time

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Pull request merged**